internal/provider/string.go

```diff
@@ -144,6 +147,20 @@ func createStringFunc(sensitive bool) func(_ context.Context, d *schema.Resource
 		minSpecial := d.Get("min_special").(int)
 		overrideSpecial := d.Get("override_special").(string)
 
+		if length < minLength {
+			return append(diags, diag.Diagnostic{
+				Severity: diag.Error,
+				Summary:  fmt.Sprintf("length (%d) must be >= %d", length, minLength),
+			})
+		}
```

Comment:
Since this validation is statically defined for a single attribute, it could be caught earlier (validation-time vs apply time) by adding a [`ValidateDiagFunc`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema#Schema.ValidateDiagFunc) to the attribute such as [`helper/validation.IntAtLeast()`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation#IntAtLeast), e.g.

```go
"length": {
	// ...
	ValidateDiagFunc: validation.ToDiagFunc(validation.IntAtLeast(1)),
},
```

Not for this pull request, but for what its worth, all this validation logic could also be performed in a [`CustomizeDiff`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema#Resource.CustomizeDiff) so it tries to run earlier at plan-time in addition to apply-time. I can walk you through that if you're interested.

Reply:
Have added usage of `ValidateDiagFunc`, thanks for the tip. Definitely interested to see `CustomizeDiff`.